src/cpr_data_access/models.py

```diff
@@ -1097,3 +1105,106 @@ def to_huggingface(
+                    text=[row["text"]],
+                    text_block_id=row["text_block_id"],
+                    language=row["language"],
+                    type=row["type"],
+                    type_confidence=row["type_confidence"],
+                    page_number=row["page_number"],
+                    coords=[tuple(c) for c in row["coords"].tolist()] if row["coords"] is not None else None,  # type: ignore
+                )
+                for _, row in doc_df.iterrows()
+            ]
+
+            # pandas to_dict() stores sequences as numpy arrays, so we need to convert them back to lists
+            doc_fields = {
+                k: list(v) if isinstance(v, np.ndarray) else v
+                for k, v in doc_df.iloc[0].to_dict().items()
+            }
+
+            doc_metadata_dict = doc_fields | {
+                "source": "GST" if self.document_model == GSTDocument else "CPR"
+            }
+
+            doc = self.document_model.parse_obj(
+                doc_fields
+                | {
+                    "document_metadata": doc_metadata_dict,
+                    "text_blocks": document_text_blocks,
+                }
+            )
+
+            documents.append(doc)
+
+        self.documents = documents
+
+        return self
+
+    def from_huggingface(
+        self,
+        dataset_name: Optional[str] = None,
+        dataset_version: Optional[str] = None,
+        limit: Optional[int] = None,
+        **kwargs,
+    ) -> "Dataset":
+        """
+        Load documents from a huggingface hub dataset.
+
+        For private repos a token should be provided either as a `token` kwarg or as
+        environment variable HUGGINGFACE_TOKEN. Any additional keyword arguments are
+        passed to the huggingface datasets load_dataset function.
+
+        :param dataset_name: name of the dataset on huggingface hub
+        :param dataset_version: version of the dataset on huggingface hub
+        :param limit: optionally limit the number of documents to load
+        :return self: with documents loaded from huggingface dataset
+        """
+
+        token = kwargs.get("token", os.getenv("HUGGINGFACE_TOKEN"))
+
+        if dataset_name is None:
+            if self.hf_hub_repo is None:
+                raise ValueError(
+                    f"Dataset name not provided and no default dataset name found for document model {self.document_model}. Provide a dataset name directly to this function."
+                )
+
+            dataset_name = self.hf_hub_repo
+            LOGGER.info(
+                f"Dataset name not provided. Using default dataset name {dataset_name} for document model {self.document_model}."
+            )
+
+        huggingface_dataset = load_dataset(
+            dataset_name, dataset_version, token=token, split="train", **kwargs
+        )
+
+        return self._from_huggingface_parquet(huggingface_dataset, limit)
```

Comment:
code itself looks great, but this file is now >1,200 lines long and is serving a few purposes! could we break it up into a couple of separate files to improve discoverability? 

with the other neighbouring files, maybe that looks something like

```
models
├── dataset.py
├── document.py
├── parser.py
└── pipeline.py
```

?

Reply:
good idea, but maybe for a separate PR? will add to the mental backlog 👍